node/staking.go

```diff
@@ -32,25 +32,79 @@ func (node *Node) UpdateStakingList(stakeInfoReturnValue *structs.StakeInfoRetur
 		return
 	}
 	node.CurrentStakes = make(map[common.Address]*structs.StakeInfo)
+	node.CurrentStakesByNode = make(map[[20]byte]*structs.StakeInfo)
+	utils.GetLogInstance().Info("UpdateStakingList",
+		"numStakeInfos", len(stakeInfoReturnValue.LockedAddresses))
 	for i, addr := range stakeInfoReturnValue.LockedAddresses {
 		blockNum := stakeInfoReturnValue.BlockNums[i]
+		blsAddr := stakeInfoReturnValue.BlsAddresses[i]
 		lockPeriodCount := stakeInfoReturnValue.LockPeriodCounts[i]
+		logger := utils.GetLogInstance().New(
+			"accountAddr", addr.Hex(),
+			"nodeAddr", hex.EncodeToString(blsAddr[:]))
 
 		startEpoch := core.GetEpochFromBlockNumber(blockNum.Uint64())
 		curEpoch := core.GetEpochFromBlockNumber(node.blockchain.CurrentBlock().NumberU64())
 
-		if startEpoch == curEpoch {
+		if startEpoch >= curEpoch {
+			logger.Info("future stake; skipping")
 			continue // The token are counted into stakes at the beginning of next epoch.
 		}
 		// True if the token is still staked within the locking period.
 		if curEpoch-startEpoch <= lockPeriodCount.Uint64()*lockPeriodInEpochs {
-			node.CurrentStakes[addr] = &structs.StakeInfo{
-				stakeInfoReturnValue.BlsAddresses[i],
+			logger.Info("active stake; recording")
+			stakeInfo := &structs.StakeInfo{
+				addr,
+				blsAddr,
 				blockNum,
 				lockPeriodCount,
 				stakeInfoReturnValue.Amounts[i],
 			}
+			node.CurrentStakes[addr] = stakeInfo
+			node.CurrentStakesByNode[blsAddr] = stakeInfo
+		} else {
+			logger.Info("expired stake; skipping")
+		}
+	}
+}
+
+// UpdateStakingListWithInitShardState updates the current stakes list with
+// initial hardcoded stakes.
+//
+// Initial stakers are assumed to be active since ever and until forever,
+// but with zero stake.
```

Comment:
what's the purpose of until forever with zero stake?

Reply:
They are the stakers hardcoded in the genesis block.  Eventually they will go away and be replaced by real ones.